lib/maverick/server/plug.ex

```diff
@@ -0,0 +1,133 @@
+defmodule Maverick.Server.Plug do
+  @behaviour Maverick.Server
+
+  def child_spec(api, opts) do
+    port = Keyword.get(opts, :port, 4000)
+
+    {Plug.Cowboy, scheme: :http, plug: api.router_module(), options: [port: port]}
+  end
+
+  def router_contents(routes) do
+    IO.inspect(routes, label: "routes")
+
+    quote location: :keep do
+      use Plug.Router
+
+      require Logger
+
+      plug(Plug.Parsers,
+        parsers: [:urlencoded, :json],
+        pass: ["text/*"],
+        json_decoder: Jason
+      )
+
+      plug(:match)
+      plug(:dispatch)
+
+      unquote(generate_match_functions(routes))
+
+      match _ do
+        var!(conn)
+        |> send_resp(404, "Not Found")
+      end
+    end
+  end
+
+  defp generate_match_functions(routes) do
+    for %Maverick.Route{
+          args: arg_type,
+          function: function,
+          method: method,
+          module: module,
+          path: path,
+          success_code: success,
+          error_code: error
+        } <-
+          routes do
+      plug_path = to_plug_path(path)
+      method_macro = method |> String.downcase() |> String.to_atom()
+
+      result =
+        quote location: :keep do
+          unquote(method_macro)(unquote(plug_path)) do
+            arg = Maverick.Server.Plug.decode_arg_type(var!(conn), unquote(arg_type))
+            response = apply(unquote(module), unquote(function), [arg])
+
+            Maverick.Server.Plug.wrap_response(
+              var!(conn),
+              response,
+              unquote(success),
+              unquote(error)
+            )
+          end
+        end
+
+      IO.puts(Macro.to_string(result))
+
+      result
+    end
+  end
+
+  defp to_plug_path(path) do
+    result =
+      Enum.map(path, fn element ->
+        case element do
+          {:variable, variable} ->
+            ":" <> variable
```

Comment:
this feels like a lot of work to undo the parser's work; would it make more sense to stuff both the raw _and_ the parsed path into the `%Maverick.Route{}` and then just have the appropriate one used by its respective handler/server?

Reply:
:+1: